eBPF Functions (enable/disable threads)

thiagoftsm opened the pull request:
##### Summary
This is the first PR bringing functions to eBPF.plugin;

##### Test Plan

1. Copy `tests/ebpf/ebpf.d.conf` to `/etc/netdata/`
2. Compile this branch with flag `NETDATA_DEV_MODE`.
3. Run next command and verify that everything succeed:

```sh
$ sh tests/ebpf/ebpf_thread_function.sh > out.txt 2> err.tx 
```

##### Additional Information
This PR was tested on :

| Distribution | Environment | Kernel | Error | Output |
|-------------------|---------------------|-----------|------------|---------|
|Slackware current | Bare metal | 6.1.36 | [slackware_6_1_err.txt](https://github.com/netdata/netdata/files/11941169/slackware_6_1_err.txt) | [slackware_6_1_out.txt](https://github.com/netdata/netdata/files/11941170/slackware_6_1_out.txt) |
| Arch Linux | Libvirt | 6.4.1-arch1-1|[arch_6_4_err.txt](https://github.com/netdata/netdata/files/11941704/arch_6_4_err.txt) | [arch_6_4_out.txt](https://github.com/netdata/netdata/files/11941705/arch_6_4_out.txt) |
|Ubuntu 22.04 | Libvirt | 5.15.0-69-generic |  [ubuntu_5_15_err.txt](https://github.com/netdata/netdata/files/11941849/ubuntu_5_15_err.txt) | [ubuntu_5_15_out.txt](https://github.com/netdata/netdata/files/11941850/ubuntu_5_15_out.txt) | 
|CentOS 9 | Libvirt | 5.14.0-331.el9.x86_64  | [centos_5_14_err.txt](https://github.com/netdata/netdata/files/11943347/centos_5_14_err.txt) | [centos_5_14_out.txt](https://github.com/netdata/netdata/files/11943348/centos_5_14_out.txt) |
| Debian 11 | Libvirt | 5.10.0-23-amd64 | [debian_5_10_err.txt](https://github.com/netdata/netdata/files/11949024/debian_5_10_err.txt) | [debian_5_10_out.txt](https://github.com/netdata/netdata/files/11949025/debian_5_10_out.txt) | 
|Ubuntu 20.04 | Libvirt | 5.4.0-146-generic | [ubuntu_5_4_err.txt](https://github.com/netdata/netdata/files/11896306/ubuntu_5_4_err.txt) | [ubuntu_5_4_out.txt](https://github.com/netdata/netdata/files/11896307/ubuntu_5_4_out.txt) | 
| Alma 8 | Libvirt | 4.18.0-477.13.1.el8_8.x86_64 | [alma_4_18_err.txt](https://github.com/netdata/netdata/files/11951013/alma_4_18_err.txt) | [alma_4_18_out.txt](https://github.com/netdata/netdata/files/11951014/alma_4_18_out.txt) |
| Ubuntu 18.04 | Libvirt | 4.15.0-208-generic | [ubuntu_4_15_err.txt](https://github.com/netdata/netdata/files/11951655/ubuntu_4_15_err.txt) | [ubuntu_4_15_out.txt](https://github.com/netdata/netdata/files/11951656/ubuntu_4_15_out.txt) | 
|Slackware current | Qemu| 4.14.290  | [slackware_4_14_err.txt](https://github.com/netdata/netdata/files/11952318/slackware_4_14_err.txt) | [slackware_4_14_out.txt](https://github.com/netdata/netdata/files/11952319/slackware_4_14_out.txt) | 
| CentOS 7 | Libvirt | 3.10.0-1160.92.1.el7.x86_64 | [centos_3_10_err.txt](https://github.com/netdata/netdata/files/11952053/centos_3_10_err.txt) | [centos_3_10_out.txt](https://github.com/netdata/netdata/files/11952054/centos_3_10_out.txt) |

When you run the script it is expected to have charts like this:

![1](https://github.com/netdata/netdata/assets/49162938/91823573-114c-4c16-b634-cc46f7bb1bcf)


<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? eBPF
- Can they see the change or is it an under the hood? If they can see it, where? no, this is an internal controller accessed via API. We are also expecting a cloud schema.
- How is the user impacted by the change?  This PR has potential to reduce overhead.
- What are there any benefits of the change?  A light plugin
</details>

MrZammler commented on collectors/ebpf.plugin/ebpf.d/hardirq.conf:
```diff
@@ -3,6 +3,9 @@
 #            new charts for the return of these functions, such as errors.
 #
-#[global]
+# The `life time` defines the time length a thread will run when is enabled by a function.
```
Ah, I hate to do this, but I think it should be `....will run when it is enabled....`

ilyam8 replied:
> The "life time" parameter specifies how long the thread will run if enabled through functions.

I am not sure "the time length" fits here.

Also, consider renaming it to `lifetime`

thiagoftsm replied:
> Ah, I hate to do this, but I think it should be ....will run when it is enabled....

In Latin languages we almost never use pronouns :facepalm:, thank you! 

@ilyam8 I updated for `lifetime`.

thiagoftsm commented on collectors/ebpf.plugin/ebpf_filesystem.c:
```diff
@@ -405,43 +413,88 @@ void ebpf_filesystem_cleanup_ebpf_data()
+            efp->hopen.title = NULL;
 
             freez(efp->hadditional.name);
+            efp->hopen.name = NULL;
```
I will never more copy and paste without change, I will never more copy and paste without change... :smile: 
Thank you!

MrZammler commented on tests/ebpf/ebpf_thread_function.sh.in:
```diff
@@ -0,0 +1,52 @@
+    fi
+
+    grep "${2}" /tmp/ebpf_netdata_test_functions.txt >/dev/null
+    if [ $TEST -ne 0 ]; then
```
Hi Thiago. Is there a `TEST=$?` missing right above this line here?

thiagoftsm replied:
Last commit fixed this, thanks!

MrZammler commented:
Hey Thiago. Some notes during testing.

First of all we need to compile with `NETDATA_DEV_MODE`.

I think there is a missing `TEST=$?` (noted above).

There is a weird delay when calling e.g. `http://localhost:19999/api/v1/function?function=ebpf_thread%20help`. Not sure why, I was looking into it. There is no such delay when calling the `processes help`. I think the delay is there also for the other endpoints.

There is a crash, when compiled with `NETDATA_INTERNAL_CHECKS`, and hits an assert:

```
#0  0x00007f7b223fd6cc in ?? () from /lib64/libc.so.6
#1  0x00007f7b223ad2b2 in raise () from /lib64/libc.so.6
#2  0x00007f7b22397472 in abort () from /lib64/libc.so.6
#3  0x00007f7b22397395 in ?? () from /lib64/libc.so.6
#4  0x00007f7b223a5ff2 in __assert_fail () from /lib64/libc.so.6
#5  0x000055e0d033e18b in _buffer_overflow_check (b=0x55e0d4bc4430) at libnetdata/buffer/../buffer/buffer.h:100
#6  buffer_increase (b=b@entry=0x55e0d4bc4430, free_size_required=free_size_required@entry=101) at libnetdata/buffer/buffer.c:281
#7  0x000055e0d03ea014 in buffer_need_bytes (needed_free_size=100, buffer=0x55e0d4bc4430) at ./libnetdata/buffer/buffer.h:142
#8  buffer_strcat (txt=0x7f7aed0f7110 "{\"status\":404,\"error_message\":\"ebpf.plugin does not have thread named dcstat\"}\n", wb=0x55e0d4bc4430) at ./libnetdata/buffer/buffer.h:224
#9  parser_action (input=0x7f7aed0f7110 "{\"status\":404,\"error_message\":\"ebpf.plugin does not have thread named dcstat\"}\n", parser=0x55e0d3065e00) at collectors/plugins.d/pluginsd_parser.h:157
#10 pluginsd_process (host=<optimized out>, cd=cd@entry=0x55e0d2fc38e0, fp_plugin_input=<optimized out>, fp_plugin_output=fp_plugin_output@entry=0x55e0d3034790, trust_durations=trust_durations@entry=0)
    at collectors/plugins.d/pluginsd_parser.c:1985
#11 0x000055e0d03de1f2 in pluginsd_worker_thread (arg=0x55e0d2fc38e0) at collectors/plugins.d/plugins_d.c:163
#12 0x000055e0d036b1c6 in netdata_thread_init (ptr=<optimized out>) at libnetdata/threads/threads.c:262
#13 0x00007f7b223fb9a3 in ?? () from /lib64/libc.so.6
#14 0x00007f7b2247dedc in ?? () from /lib64/libc.so.6
```

I've bypassed this, and the tests seem to run, but eventually I stumbled across another crash related to it, during shutdown (not always, only got it once):

```
#0  0x00005573cc7d7368 in buffer_strcat (txt=0x7f572da62110 "{\"status\":404,\"error_message\":\"ebpf.plugin does not have thread named dcstat\"}\n", wb=0x5573d075bc80)
    at ./libnetdata/buffer/buffer.h:230
230	           *d++ = *t++;
[Current thread is 1 (Thread 0x7f572da6a6c0 (LWP 5277))]
(gdb) where
#0  0x00005573cc7d7368 in buffer_strcat (txt=0x7f572da62110 "{\"status\":404,\"error_message\":\"ebpf.plugin does not have thread named dcstat\"}\n", wb=0x5573d075bc80)
    at ./libnetdata/buffer/buffer.h:230
#1  parser_action (input=0x7f572da62110 "{\"status\":404,\"error_message\":\"ebpf.plugin does not have thread named dcstat\"}\n", parser=0x5573cf42af40)
    at collectors/plugins.d/pluginsd_parser.h:157
#2  pluginsd_process (host=<optimized out>, cd=cd@entry=0x5573cf3e1ad0, fp_plugin_input=<optimized out>, fp_plugin_output=fp_plugin_output@entry=0x5573cf428630, 
    trust_durations=trust_durations@entry=0) at collectors/plugins.d/pluginsd_parser.c:1985
#3  0x00005573cc7cc587 in pluginsd_worker_thread (arg=0x5573cf3e1ad0) at collectors/plugins.d/plugins_d.c:163
#4  0x00005573cc76a786 in netdata_thread_init (ptr=<optimized out>) at libnetdata/threads/threads.c:262
#5  0x00007f574652d963 in ?? () from /lib64/libc.so.6
#6  0x00007f57465afe9c in ?? () from /lib64/libc.so.6
```

I'll update here if I find something related to the above.

MrZammler commented on collectors/ebpf.plugin/ebpf.c:
```diff
@@ -2373,6 +2372,278 @@ static void ebpf_parse_args(int argc, char **argv)
+    EBPF_PLUGIN_FUNCTIONS(EBPF_FUNCTION_THREAD, EBPF_PLUGIN_THREAD_FUNCTION_DESCRIPTION);
+
+    ebpf_create_lifetime_thread_chart(update_every);
+    EBPF_PLUGIN_FUNCTIONS(EBPF_FUNCTION_THREAD, EBPF_PLUGIN_THREAD_FUNCTION_DESCRIPTION);
```
Thiago, just a note, is it ok to send `EBPF_PLUGIN_FUNCTIONS(EBPF_FUNCTION_THREAD, EBPF_PLUGIN_THREAD_FUNCTION_DESCRIPTION);` twice here?

thiagoftsm replied:
This is correct. I am adding them for charts associated with the function. We have the same pattern in [apps.plugn](https://github.com/netdata/netdata/blob/b51c4278dd251c45daba5d2f457755da936651e3/collectors/apps.plugin/apps_plugin.c#L3780-L3799)

thiagoftsm commented:
> Hey Thiago. Some notes during testing.
> 
> First of all we need to compile with `NETDATA_DEV_MODE`.

I updated the OP, thank you!

> I think there is a missing `TEST=$?` (noted above).
I fixed this with a commit.

> There is a weird delay when calling e.g. `http://localhost:19999/api/v1/function?function=ebpf_thread%20help`. Not sure why, I was looking into it. There is no such delay when calling the `processes help`. I think the delay is there also for the other endpoints.

I am taking a look on this right now.

> There is a crash, when compiled with `NETDATA_INTERNAL_CHECKS`, and hits an assert:
I will try to recreate this, but  I did not modify our core functions associated with this call. :thinking: 
> 
> I've bypassed this, and the tests seem to run, but eventually I stumbled across another crash related to it, during shutdown (not always, only got it once):

I will also take a look on this, thank you!

> I'll update here if I find something related to the above.

Great! :)

thiagoftsm commented:
@MrZammler and @stelfrag I ran valgrind to detect issues in eBPF code and it does not detect, but we have some reports for our `buffer_create` functions, please, take a look in [this](https://github.com/netdata/netdata/files/11923322/valgrind.txt) report.

MrZammler commented:
Thiago hi! Is it possible to rebase the PR after https://github.com/netdata/netdata/pull/15293 ? Should fix the delay issue on calling the function (and make testing faster). Thanks!

thiagoftsm commented:
> Thiago hi! Is it possible to rebase the PR after #15293 ? Should fix the delay issue on calling the function (and make testing faster). Thanks!

Hello @MrZammler ,

I rebased and the issue was fixed :partying_face: , I am retesting on all environments again, please, take another look.

Best regards!

Ancairon commented on collectors/ebpf.plugin/README.md:
```diff
@@ -989,3 +989,50 @@ shows how the lockdown module impacts `ebpf.plugin` based on the selected option
+
+The eBPF plugin has a [function](https://github.com/netdata/netdata/blob/master/docs/cloud/netdata-functions.md) named
+`ebpf_thread` that controls its internal threads and helps to reduce the overhead on host. Using the function you
+can run eBPF plugin with all threads disabled and enable them only when to take a look in specific areas is necessary.
```
```suggestion
can run the plugin with all threads disabled and enable them only when you want to take a look in specific areas.
```

Ancairon commented on collectors/ebpf.plugin/README.md:
```diff
@@ -989,3 +989,50 @@ shows how the lockdown module impacts `ebpf.plugin` based on the selected option
+
+### Disable thread
+
+It is also possible to stop any thread running using keywork `disable`, for exaample, to disable `cachestat` you can
```
```suggestion
It is also possible to stop any thread running using the keyword `disable`. For example, to disable `cachestat` you can
```

Ancairon commented on collectors/ebpf.plugin/README.md:
```diff
@@ -989,3 +989,50 @@ shows how the lockdown module impacts `ebpf.plugin` based on the selected option
+
+### Debugging threads
+
+You can verify the impact of threads on host running the
```
```suggestion
You can verify the impact of threads on the host by running the
```

Ancairon commented on collectors/ebpf.plugin/README.md:
```diff
@@ -989,3 +989,50 @@ shows how the lockdown module impacts `ebpf.plugin` based on the selected option
+[ebpf_thread_function.sh](https://github.com/netdata/netdata/blob/master/tests/ebpf/ebpf_thread_function.sh)
+script on your environment.
+
+You can check the results of to have threads running on your environment in Netdata monitoring section on your
```
```suggestion
You can check the results of having threads running on your environment in the Netdata monitoring section on your
```

Ancairon commented on collectors/ebpf.plugin/README.md:
```diff
@@ -989,3 +989,50 @@ shows how the lockdown module impacts `ebpf.plugin` based on the selected option
+
+`http://localhost:19999/api/v1/function?function=ebpf_thread%20enable:mount`
+
+this will run thread `mount` during 300 seconds (5 minutes). You can specify a specific period appending the period
```
```suggestion
this will run thread `mount` during 300 seconds (5 minutes). You can specify a specific period by appending the period
```

MrZammler commented on collectors/ebpf.plugin/ebpf_functions.c:
```diff
@@ -0,0 +1,413 @@
+                else
+                    ebpf_function_error(transaction,
+                                        HTTP_RESP_NOT_FOUND,
+                                        "No function with this name found in ebpf.plugin.");
```
Hi Thiago!

Seems adding a `fflush(stdio);` here, seems to solve my issues with delay of getting `http://localhost:19999/api/v1/function?function=ebpf_thread%20help` (6ms vs 3-4 secs). Could you try as well?

thiagoftsm replied:
Thank you @MrZammler , I pushed a commit adding not only the `fflush`, but also the necessary `mutex` around.

MrZammler commented:
I've re-run the tests on a Debian 11 VM and appear ok! (Sadly I still get random issues on my Gentoo).

MrZammler commented on tests/ebpf/ebpf_thread_function.sh.in:
```diff
@@ -0,0 +1,50 @@
+    echo "TESTING ${THREAD}"
+    netdata_ebpf_test_functions "${MURL}/api/v1/function?function=ebpf_thread%20enable:${THREAD}:${INTERVAL}%20thread:${THREAD}"
+    sleep 17
+    netdata_ebpf_test_functions "${MURL}/api/v1/function?function=ebpf_thread%20thread:${THREAD}" "running"
```
Also a small issue on the test, `running` will exist in the output, since json will always have: `"help":"Detailed information on the currently running threads.",` so it could grep that. Maybe adding a more strict grep?

thiagoftsm replied:
I modified the help sentence to be more generic, so this grep can stay like it is.

thiagoftsm replied:
Thank you! :)

MrZammler commented:
Also, for some weird reason, when I request: `curl 'http://localhost:19999/api/v1/function?function=ebpf_thread%20thread:shm'` I get back help for `"socket","Monitors TCP and UDP bandwidth. This thread is integrated with apps and cgroup.","stopped",0,"NULL"]]` could you check please?

thiagoftsm commented:
> I've re-run the tests on a Debian 11 VM and appear ok! (Sadly I still get random issues on my Gentoo).

Gentoo issues will be fixed when the PRs https://github.com/netdata/netdata/pull/15070  and https://github.com/netdata/netdata/pull/15323 are merged.

thiagoftsm commented:
> Also, for some weird reason, when I request: `curl 'http://localhost:19999/api/v1/function?function=ebpf_thread%20thread:shm'` I get back help for `"socket","Monitors TCP and UDP bandwidth. This thread is integrated with apps and cgroup.","stopped",0,"NULL"]]` could you check please?

As I said in the meeting this morning, I am starting to work with sockets for we have a new feature based in this thread. We have a lot of work to do in this specific thread, so I am removing it from current tests and I will bring back after to reorganize the thread.

thiagoftsm commented:
> Edit: Not entirelly sure why this happens, it appears to happen when:
>
> A thread is stopped, I enable it, then I query it again:

Last commit fixed this issue you found during your tests.

MrZammler commented:
> > Edit: Not entirelly sure why this happens, it appears to happen when:
> > A thread is stopped, I enable it, then I query it again:
> 
> Last commit fixed this issue you found during your tests.

Indeed, seems to work much better now!

MrZammler commented:
Hi Thiago. On one of the tests, where I had `shm` running then turned it off, I got:

```
cm@manolis-vm:~$ curl 'http://localhost:19999/api/v1/function?function=ebpf_thread%20disable:shm'
CHART apps.shmget_call '' 'Calls to syscall <code>shmget(2)</code>.' 'calls/s' 'ipc shm (eBPF)' '' 'stacked' 20191 5 'obsolete'
CHART apps.shmat_call '' 'Calls to syscall <code>shmat(2)</code>.' 'calls/s' 'ipc shm (eBPF)' '' 'stacked' 20192 5 'obsolete'
CHART apps.shmdt_call '' 'Calls to syscall <code>shmdt(2)</code>.' 'calls/s' 'ipc shm (eBPF)' '' 'stacked' 20193 5 'obsolete'
CHART apps.shmctl_call '' 'Calls to syscall <code>shmctl(2)</code>.' 'calls/s' 'ipc shm (eBPF)' '' 'stacked' 20194 5 'obsolete'
CHART system.shared_memory_calls '' 'Calls to shared memory system calls' 'calls/s' 'ipc shared memory' '' 'line' 1207 5 'obsolete'
CHART netdata.aral_ebpf_shm_alloc '' 'Calls to allocate memory.' 'calls' 'ebpf' 'netdata.ebpf_aral_stat_alloc' 'stacked' 140100 5 'obsolete'
CHART netdata.aral_ebpf_shm_alloc '' 'Calls to allocate memory.' 'calls' 'ebpf' 'netdata.ebpf_aral_stat_alloc' 'stacked' 140101 5 'obsolete'
{
    "status":200,
    "type":"table",
    "update_every":5,
    "help":"Detailed information about eBPF threads.",
    "data":[
        ["process","Monitor information about process life. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["socket","Monitors TCP and UDP bandwidth. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["cachestat","Monitor Linux page cache internal functions. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["sync","Monitor calls to syscalls sync(2), fsync(2), fdatasync(2), syncfs(2), msync(2), and sync_file_range(2).","stopped",0,"NULL"],
        ["dc","Monitor file access using directory cache. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["swap","Monitor swap space usage. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["vfs","Monitor VFS (Virtual File System) functions. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["filesystem","Monitor filesystem latency for: btrfs, ext4, nfs, xfs and zfs.","stopped",0,"NULL"],
        ["disk","Monitor disk latency independent of filesystem.","stopped",0,"NULL"],
        ["mount","Show calls to syscalls mount(2) and umount(2).","stopped",0,"NULL"],
        ["fd","Monitor when files are open and closed. This thread is integrated with apps and cgroup.","stopped",0,"NULL"],
        ["hardirq","Show time spent servicing individual hardware interrupt requests (hard IRQs).","stopped",0,"NULL"],
        ["softirq","Show time spent servicing individual software interrupt requests (soft IRQs).","stopped",0,"NULL
```

It seems I got some `CHART` commands in stdout, not sure why....

I can't replicate it again, seems to have happened once. Maybe some lock?

thiagoftsm commented:
> I can't replicate it again, seems to have happened once. Maybe some lock?

I never saw this, but it looks like we still had stdout buffered, because `shm` charts did not fill 4k bytes with data. I will push a commit to address this.
